File: plenopy/photon_stream/reader.py
import numpy as np


def py_stream2sequence(
    photon_stream,
    time_slice_duration,
    NEXT_READOUT_CHANNEL_MARKER,
    sequence,
    time_delay_mean
):
    number_time_slices = sequence.shape[0]

    lixel = 0
    for symbol in photon_stream:
        if symbol == NEXT_READOUT_CHANNEL_MARKER:
            lixel += 1
        else:
            time_slice = symbol

            arrival_time = time_slice*time_slice_duration
            arrival_time -= time_delay_mean[lixel]

            arrival_slice = int(
                np.round(
                    arrival_time/time_slice_duration))

            if arrival_slice < number_time_slices and arrival_slice >= 0:
                sequence[arrival_slice, lixel] += 1.0
            else:
                pass


def py_photon_stream_to_image_sequence(
    photon_stream,
    photon_stream_next_channel_marker,
    time_slice_duration,
    time_delay_image_mean,
    projection_links,
    projection_starts,
    projection_lengths,
    number_lixel,
    number_pixel,
    number_time_slices,
):
    photon_stream_length = photon_stream.shape[0]

    assert time_slice_duration >= 0.0
    assert number_lixel == projection_starts.shape[0]
    assert number_lixel == projection_lengths.shape[0]
    assert number_lixel == time_delay_image_mean.shape[0]

    out_image_sequence = np.zeros(
        shape=(number_time_slices, number_pixel),
        dtype=np.uint32
    )

    phs_lixel = 0
    for phs_i in range(photon_stream_length):
        phs_symbol = photon_stream[phs_i]

        if phs_symbol == photon_stream_next_channel_marker:
            phs_lixel += 1
        else:
            raw_arrival_time_slice = phs_symbol

            arrival_time = raw_arrival_time_slice*time_slice_duration
            arrival_time -= time_delay_image_mean[phs_lixel]

            arrival_slice = int(np.round(arrival_time/time_slice_duration))

            if arrival_slice < number_time_slices and arrival_slice >= 0:

                for p in range(projection_lengths[phs_lixel]):
                    pp = projection_starts[phs_lixel] + p
                    pixel = projection_links[pp]
                    out_image_sequence[arrival_slice, pixel] += 1

    return out_image_sequence

File: plenopy/photon_stream/reader.c
#include <math.h>


void c_stream2sequence(
    unsigned char* photon_stream,
    unsigned int photon_stream_length,
    unsigned char NEXT_READOUT_CHANNEL_MARKER,
    unsigned short* sequence,
    unsigned int number_time_slices,
    unsigned int number_lixel,
    float *time_delay_mean,
    float time_slice_duration
) {
    unsigned int lixel = 0;
    for(unsigned int s=0; s<photon_stream_length; s++) {
        if(photon_stream[s] == NEXT_READOUT_CHANNEL_MARKER) {
            lixel++;
        }else{
            unsigned char raw_arrival_slice = photon_stream[s];
            float arrival_time = (float)raw_arrival_slice*time_slice_duration;
            arrival_time -= time_delay_mean[lixel];

            int arrival_slice = (int)round(arrival_time/time_slice_duration);

            if(arrival_slice < (int)number_time_slices && arrival_slice >= 0) {
                unsigned int index = arrival_slice*number_lixel + lixel;
                sequence[index] += 1;
            }
        }
    }

    return;
}

void c_stream2_cx_cy_arrivaltime_point_cloud(
    unsigned char* photon_stream,
    unsigned int photon_stream_length,
    unsigned char NEXT_READOUT_CHANNEL_MARKER,
    float* point_cloud,
    float* cx,
    float* cy,
    float* time_delay_mean,
    float time_slice_duration,
    unsigned int* lixel_ids
) {
    unsigned int lixel = 0;
    unsigned int photon = 0;
    for (unsigned int s = 0; s < photon_stream_length; s++) {
        if (photon_stream[s] == NEXT_READOUT_CHANNEL_MARKER) {
            lixel++;
        } else {
            unsigned char raw_arrival_slice = photon_stream[s];
            float arrival_time = (float)raw_arrival_slice*time_slice_duration;
            arrival_time -= time_delay_mean[lixel];

            point_cloud[photon*3 + 0] = cx[lixel];
            point_cloud[photon*3 + 1] = cy[lixel];
            point_cloud[photon*3 + 2] = arrival_time;
            lixel_ids[photon] = lixel;
            photon++;
        }
    }
    return;
}

void c_photon_stream_to_arrival_slices_and_lixel_ids(
    unsigned char* photon_stream,
    unsigned int photon_stream_length,
    unsigned char NEXT_READOUT_CHANNEL_MARKER,
    unsigned char* arrival_slices,
    unsigned int* lixel_ids
) {
    unsigned int lixel = 0;
    unsigned int photon = 0;
    for (unsigned int s = 0; s < photon_stream_length; s++) {
        if (photon_stream[s] == NEXT_READOUT_CHANNEL_MARKER) {
            lixel++;
        } else {
            arrival_slices[photon] = photon_stream[s];
            lixel_ids[photon] = lixel;
            photon++;
        }
    }
    return;
}


void c_photon_stream_to_image_sequence(
    unsigned char* photon_stream,
    unsigned int photon_stream_length,
    unsigned char photon_stream_next_channel_marker,
    float time_slice_duration,
    float *time_delay_image_mean,

    unsigned int *projection_links,
    unsigned int *projection_starts,
    unsigned int *projection_lengths,

    unsigned int number_lixel,
    unsigned int number_pixel,
    unsigned int number_time_slices,

    unsigned int *image_sequence)
{
    unsigned int lixel = 0;
    unsigned int s = 0;

    for (s = 0; s < photon_stream_length; s++) {
        if (photon_stream[s] == photon_stream_next_channel_marker) {
            lixel++;
        } else {
            unsigned char raw_arrival_slice = photon_stream[s];
            float arrival_time = (float)raw_arrival_slice*time_slice_duration;
            arrival_time -= time_delay_image_mean[lixel];
            int arrival_slice = (int)round(arrival_time/time_slice_duration);
            if(arrival_slice < (int)number_time_slices && arrival_slice >= 0) {
                unsigned int p = 0;
                unsigned int p_start = projection_starts[lixel];
                unsigned int p_stop = p_start + projection_lengths[lixel];
                for (p = p_start; p < p_stop; p++) {
                    unsigned int pixel = projection_links[p];
                    unsigned int idx = arrival_slice*number_pixel + pixel;
                    image_sequence[idx] += 1;
                }
            }
        }
    }
}
